File: linux/device/go/internal/echo/service.go
// SPDX-License-Identifier: AGPL-3.0-only

// Package echo is a smoke-test app: it registers over IPC and answers "ping"
// commands with "pong". It demonstrates the entire app authoring surface —
// connect, auto-register, dispatch commands, respond — in ~40 lines.
package echo

import (
	"context"
	"log/slog"

	"github.com/helix-kit/helix-device/internal/echo/generated"
	"github.com/helix-kit/helix-device/internal/ipc"
	"github.com/helix-kit/helix-device/internal/shared/config"
	"github.com/helix-kit/helix-device/internal/shared/ipcutil"
	"github.com/helix-kit/helix-device/internal/shared/servicemain"
)

// ServiceName is how this app registers with helixd (its `message.service`).
const ServiceName = generated.EchoService

type runner struct {
	cfg    *config.Config
	log    *slog.Logger
	client *ipc.Client
}

// New builds the echo Runner.
func New(cfg *config.Config, log *slog.Logger) (servicemain.Runner, error) {
	return &runner{cfg: cfg, log: log}, nil
}

func (r *runner) Run(ctx context.Context) error {
	r.client = ipc.NewClient(r.cfg.IPC.SocketPath, ServiceName, r.log)
	r.client.Start()
	if err := r.client.WaitUntilConnected(ctx); err != nil {
		return err
	}
	go ipcutil.DispatchCommands(ctx, r.client.Notifications(), r.handle)
	<-ctx.Done()
	return nil
}

func (r *runner) Close() {
	if r.client != nil {
		r.client.Close()
	}
}

func (r *runner) handle(ctx context.Context, cmd ipc.CommandParams) {
	generated.DispatchEcho(ctx, r.client, r.log, cmd, r)
}

// HandlePing answers a ping command with the echoed message and this service name.
func (r *runner) HandlePing(
	_ context.Context,
	req generated.EchoPingInput,
) (generated.EchoPongOutput, error) {
	return generated.EchoPongOutput{Echo: req.Message, Service: ServiceName}, nil
}

File: linux/device/python/helix_echo/src/helix_echo/service.py
# SPDX-License-Identifier: AGPL-3.0-only
"""Echo service: register over IPC and answer "ping" commands with "pong"."""

from __future__ import annotations

import logging
import threading

from helix_service_runtime import (
    DeviceConfig,
    IPCClient,
    Service,
    run_command_loop,
    run_service_main,
)

from .generated.echo import (
    ECHO_SERVICE,
    EchoPingInput,
    EchoPongOutput,
    dispatch_echo,
)


class EchoService:
    """Implements the generated ``EchoHandler`` and pumps IPC commands into it."""

    def __init__(self, config: DeviceConfig, log: logging.Logger) -> None:
        self._log = log
        self._ipc = IPCClient(config.ipc_socket_path, ECHO_SERVICE, log)
        self._stop = threading.Event()

    def handle_ping(self, request: EchoPingInput) -> EchoPongOutput:
        return EchoPongOutput(echo=request.message or "", service=ECHO_SERVICE)

    def run_forever(self) -> None:
        self._ipc.connect()
        try:
            run_command_loop(
                self._ipc,
                self._stop,
                lambda command: dispatch_echo(self._ipc, self._log, command, self),
            )
        finally:
            self._ipc.close()


def build_service(config: DeviceConfig, log: logging.Logger) -> Service:
    return EchoService(config, log)


def main() -> int:
    return run_service_main(
        None,
        description="Helix Python echo device app",
        version="0.1.0",
        logger_name="helix-echo",
        build_service=build_service,
    )
